### ISSUE TITLE
closeCurrentWindow() doesn't work headless

### ISSUE THREAD

**Issue by ribosradagast** [TestFX/TestFX#447]
I have noticed that closeCurrentWindow() doesn't actually do anything when running headless...  is this expected?  Is there a workaround?

**Comment by ribosradagast** [comment]
It's easy enough to test, too.  Make a test that makes an 
`Alert alert = new Alert(AlertType.ERROR);
alert.showAndWait();`

then in the test, call the method that shows the error

then you can 
`verifyThat(".error", Node::isVisible); 
closeCurrentWindow();
if (lookup(".error").tryQuery().isPresent()) {
   Assert.fail("Window was not closed");
}`

this test works non-headless but fails in headless mode.

**Comment by ribosradagast** [comment]
This is on Windows.  I did look into the source code just like you did and confirmed that Alt-F4 doesn't do anything in a headless context

**Comment by brcolow** [comment]
Can you try to grab the window/stage using FxRobot.listWindows(), FxRobot.listTargetWindows(), or any of the FxRobot.window(..) methods? Then request close on that?

**Comment by ribosradagast** [comment]
I actually got a reference to the stage by other means and called .close() on it there too... no dice.

Also tried just doing stage.getOnCloseRequest().handle(<a made up stage closing object> ) and that didn't seem to help either

**Comment by brcolow** [comment]
Hmm what does do when you run your test:

 `stage.getOwner().setOnHidden(() -> System.out.println("Window has been hidden!");`

Also can you check what stage.getOwner().isShowing() returns after you expect it to be closed?

**Comment by ribosradagast** [comment]
hmm stage.getOwner() is giving me null.  However, doing stage.setOnHidden () yields me noprintout in the headless run

**Comment by brcolow** [comment]
`getOwner()` will only work if you use `initOwner(owner)` before making the `Dialog` (in this case, `Alert`) visible. I am not sure if it will make a difference in this case, though. Also does calling alt-f4 on an `Alert` in headed mode work correctly (i.e. close the `Alert`)? I am not sure what the original intention/usage of the `closeCurrentWindow()` method was as no test was added that uses it.

**Comment by ribosradagast** [comment]
Yeah it doesn't seem to make a difference from what I can tell.  Alt-F4 works in non-headless mode just fine.

How would you suggest that we close out of errors in our unit tests, if calling .close() is unreliable in headless?  Thanks

**Comment by brcolow** [comment]
I thought that you are using `closeCurrentWindow()` not calling `close()` directly on the Alert?

**Comment by ribosradagast** [comment]
No sorry you're right, calling .close() directly on the alert does work.  However, in application we don't have a handle directly to that Stage, and it seems like a long shot to get one.    Here's a SSCCE

**Comment by ribosradagast** [comment]
```
public class CloseWindowTest extends ApplicationTest {
	String ERROR_TITLE = "Error Window TItle";
Button dockButton;
BorderPane pane;
static Alert alert;

public static void showErrorWindow(String string) {
Platform.runLater(() -> {
alert = new Alert (Alert.AlertType.ERROR);
alert.setTitle(string);
alert.setHeader(string);
alert.setContent(string);
alert.showAndWait();
});
}

@Test
public void testClosingCurrentWindow() throws InterruptedException {
interact(() -> showErrorWindow(ERROR_TITLE));

verifyThat(".error", Node::isVisible);

//close the error window
closeCurrentWindow();

//there should be no more error window
try {
targetWindow(ERROR_TITLE);
Assert.fail("Should have thrown an exception - window shouldn't exist anymore");
}catch(NoSuchElementException ns) {
// this was expected.  continue.
}
@Override
public void init () throws TimeoutException {
FxToolkit.registerStage(() -> new Stage());
}

@Override 
public void start (Stage stage) throws InterruptedException {
Label leftLabel = new Label("Left label");
dockButton = new Button ("Test button");

pane = new BorderPane();
pane.setLeft(leftLabel);
pane.setCenter(dockButton);

stage.setScene(new Scene(pane));

stage.show();
}

@Override
public void stop() throws TimeoutException {
FXToolkit.hideStage();
}
}



```

**Comment by brcolow** [comment]
Even it is a long shot to somehow get a handle to the alert stage (can you call `setId()` on it and access via `lookup(#id)`?) try and get one and call `close()` on it directly as a /workaround/ to this issue. I will try and fix `closeCurrentWindow()` but in the meantime if you can find a workaround that would be swell. Thanks for the report and your patience.

**Comment by ribosradagast** [comment]
Future people out there finding this through google, here's how to close a Dialog or Alert through TestFX:
```java
interact(()->((Stage)((lookup(".error").query())).getScene().getWindow()).close());
```
just got to hope to goodness that none of that is null xD
